cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/LogPipelineConfigurator.java

```diff
@@ -98,12 +102,78 @@ public void begin(InterpretationContext ec, String name, Attributes attributes)
       Map<String, Appender<E>> appenderBag = (Map<String, Appender<E>>) ec.getObjectMap().get(ActionConst.APPENDER_BAG);
 
       Appender<E> appender = appenderBag.get(appenderName);
-      appenderBag.put(appenderName, new FlushableAppender<>(appender));
+      appenderBag.put(appenderName, new WrappedAppender<>(appender));
     }
 
     @Override
     public void end(InterpretationContext ec, String name) throws ActionException {
       // no-op
     }
   }
+
+  /**
+   * An {@link Appender} that implements both {@link Flushable} and {@link Syncable}. If either
+   * {@link #flush()} or {@link #sync()} failed, it will be retried on the next {@link #doAppend(Object)}
+   * call.
+   *
+   * @param <E> type of event that can be appended.
+   */
+  private static final class WrappedAppender<E> extends ForwardingAppender<E> implements Flushable, Syncable {
+
+    private boolean needFlush;
+    private boolean needSync;
+
+    WrappedAppender(Appender<E> delegate) {
+      super(delegate);
+    }
+
+    @Override
+    public void doAppend(E event) throws LogbackException {
+      if (needSync) {
+        try {
+          sync();
```

Comment:
no, when?? This is the retry mechanism such that if last call to sync failed, it will get retried on the next `doAppend` call. If the retry succeeded, we'll go ahead and append